In [1]:
# Importing the libraries for data analysis
import pandas as pd


In [2]:
# Install xlrd to read excel files
#!pip install xlrd

In [3]:
#Import the file with the functions
import functions as f

In [4]:
# URL of the Excel file containing shark attack data
url = 'https://www.sharkattackfile.net/spreadsheets/GSAF5.xls'  


In [5]:
# Read the Excel file from the given URL into a DataFrame
#df = pd.read_excel(url)
df = f.read_data(url)

# Create a copy of the DataFrame to make changes
df_mod = df.copy()

# Now you can work with df_mod without affecting the original df

In [6]:
# Display the first few rows of the DataFrame
df_mod.head()

Date    Year        Type    Country              State  \
0  2024-09-16 00:00:00  2024.0  Unprovoked    Morocco   Southern Morocco   
1  2024-08-26 00:00:00  2024.0  Unprovoked    Jamaica        Montego Bay   
2  2024-08-06 00:00:00  2024.0  Unprovoked     Belize   Gulf of Honduras   
3  2024-07-23 00:00:00  2024.0  Unprovoked  Australia                NSW   
4  2024-07-18 00:00:00  2024.0  Unprovoked  Australia  Western Australia   

                            Location                     Activity  \
0                     West of Dakhla  Swimming - jumped off yacht   
1                           Falmouth                 Spearfishing   
2                    Central America           Swimming - Diving    
3  North Shore Beach, Port Macquarie                      Surfing   
4               Trigg beach Sterling                      Surfing   

                Name Sex Age  ...                        Species   \
0     German Tourist   F  30  ...        Reportedly a Great White   
1       Jahmari Reid   M  16  ...          Reportedly Tiger Shark   
2  Annabelle Carlson   F  15  ...           Reef shark 6ft (1.8m)   
3       Kai McKenzie   M  23  ...                  3m White Shark   
4       Ryan Lowther   M  41  ...  Undetermined small shall shark   

                                              Source  pdf href formula href  \
0         Andy Currie: Moroccan World News: Sky News  NaN          NaN  NaN   
1   Todd Smith: Daily Mail UK: Sky News: People .com  NaN          NaN  NaN   
2  Kevin McMurray Tracking Sharks.com New York Po...  NaN          NaN  NaN   
3          Nine.com.au: ABC.Net.Au Johannes Marchand  NaN          NaN  NaN   
4          Daily Mail: Sky News: The West Australian  NaN          NaN  NaN   

  Case Number Case Number.1 original order Unnamed: 21 Unnamed: 22  
0         NaN           NaN            NaN         NaN         NaN  
1         NaN           NaN            NaN         NaN         NaN  
2         NaN           NaN            NaN         NaN         NaN  
3         NaN           NaN            NaN         NaN         NaN  
4         NaN           NaN            NaN         NaN         NaN  

[5 rows x 23 columns]

In [7]:
# Show the shape of the DataFrame
print(f"The DataFrame has {df_mod.shape[0]} rows and {df_mod.shape[1]} columns.")

The DataFrame has 6970 rows and 23 columns.


In [8]:
# Display the information of the DataFrame
df_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  pdf             6799 non-null   object 
 16  href formula    6794 non-null   object 
 17  href            6796 non-null   o

In [9]:
# Check for missing values in the DataFrame
df_mod.isnull().sum()

Date                 0
Year                 2
Type                18
Country             50
State              482
Location           565
Activity           586
Name               220
Sex                579
Age               2995
Injury              35
Unnamed: 11        562
Time              3527
Species           3132
Source              19
pdf                171
href formula       176
href               174
Case Number        172
Case Number.1      173
original order     171
Unnamed: 21       6969
Unnamed: 22       6968
dtype: int64

## Drop the spaces in Columns Names

In [10]:
# Remove leading and trailing spaces from column names
df_mod.columns = df_mod.columns.str.strip()

In [11]:
# Display the column names of the modified DataFrame
df_mod.columns

Index(['Date', 'Year', 'Type', 'Country', 'State', 'Location', 'Activity',
       'Name', 'Sex', 'Age', 'Injury', 'Unnamed: 11', 'Time', 'Species',
       'Source', 'pdf', 'href formula', 'href', 'Case Number', 'Case Number.1',
       'original order', 'Unnamed: 21', 'Unnamed: 22'],
      dtype='object')

## Drop the Useless Columns

In [12]:
# Display the column names of the modified DataFrame
df_mod.columns

Index(['Date', 'Year', 'Type', 'Country', 'State', 'Location', 'Activity',
       'Name', 'Sex', 'Age', 'Injury', 'Unnamed: 11', 'Time', 'Species',
       'Source', 'pdf', 'href formula', 'href', 'Case Number', 'Case Number.1',
       'original order', 'Unnamed: 21', 'Unnamed: 22'],
      dtype='object')

In [13]:
# Drop unnecessary columns from the DataFrame
df_mod = f.drop_columns(df_mod)


In [14]:
# Display the column names of the modified DataFrame
df_mod.columns

Index(['Date', 'Year', 'Type', 'Country', 'State', 'Location', 'Activity',
       'Name', 'Sex', 'Age', 'Injury', 'Unnamed: 11', 'Time', 'Species',
       'Source', 'original order'],
      dtype='object')

## Analysis of Duplicated Rows

In [15]:
# Count and print the number of duplicated rows in the DataFrame
num_dupl = df_mod.duplicated().sum()
print(f"There are {num_dupl} duplicated rows")

There are 1 duplicated rows


In [16]:
# Step 1: Eliminate duplicated row

df_mod.drop_duplicates(keep='first', inplace=True)

# Step 2: reset index
df_mod.reset_index(drop=True, inplace=True)

# Step 3: Check for duplicated rows
num_dupl = df_mod.duplicated().sum()
print(f"There are {num_dupl} duplicated rows")

There are 0 duplicated rows



## Analysis of DataFrame Columns

In this section, we will begin the analysis of the columns in the DataFrame. We will examine each column to understand its data, identify any necessary transformations, and clean the data as required. This process will help us prepare the data for further analysis and visualization.


### Analysis of the Date Column

In this section, we will analyze the `Date` column of the DataFrame. First, we will convert the `Date` column to a date data type to facilitate its manipulation and analysis. Then, we will explore the column to obtain information such as the date range and the frequency of events by year, month, etc.


In [17]:
# Show all the unique values in the 'Date' column
for date in df_mod['Date'].unique():
    print(date)

2024-09-16 00:00:00
2024-08-26 00:00:00
2024-08-06 00:00:00
2024-07-23 00:00:00
2024-07-18 00:00:00
2024-07-08 00:00:00
2024-07-05 00:00:00
2024-07-04 00:00:00
2024-06-07 00:00:00
2024-06-02 00:00:00
2024-05-28 00:00:00
2024-05-24 00:00:00
2024-05-23 00:00:00
2024-05-02 00:00:00
2024-04-26 00:00:00
2024-04-20 00:00:00
2024-04-08 00:00:00
15 Mar 2024
04 Mar 2024
02 Mar-2024
25 Feb-2024
14 Feb-2024
04-Feb-2024
29 Jan-2024
15 Jan-2024
09-Jan-2024
05-Jan-2024
30 Dec-2023
29 Dec-2023
28 Dec-2023
25 Dec-2023
24-Dec-2023
18 Dec-2023
14  Dec-2023
08 Dec-2023
04 Dec-2023
02 Dec-2023
30 Nov-2023
21 Nov-2023
10 Nov-2023
05 Nov-2023
03 Nov-2023
02 Nov-2023
Reported 02 Nov-2023
31 Oct-2023
25 Oct 2023
23 Oct-2023
21 Oct-2023
20 Oct 2023
15 Oct 2023
13 Oct-2023
02 Oct-2023
01 Oct-2023
18 Sep-2023
13 Sep-2023
12 Sep-2023
11 Sep-2023
09 Sep- 2023
Reported 06-Sep-2023
04 Sep 2023
25 Aug 2023
23 Aug-2023
22 Aug-2023
21 Aug-2023
16 Aug-2023
12 Aug-2023
07 Aug-2023
31 Jul-2023
30 Jul-2023
29 Jul-2023
24 J

In [18]:
# Step 1: Clean data
df_mod = f.date_clean(df_mod)
#df_mod['Date'].head()

In [19]:
# Step 2: Convert to Datetime and Verify Null Values
df_mod['Date'] = pd.to_datetime(df_mod['Date'], errors='coerce')
null_dates_count = df_mod['Date'].isnull().sum()
print(f"Number of null dates after conversion: {null_dates_count}")

Number of null dates after conversion: 415


In [20]:
# Step 3: Handle Null Values
# Option 1: Drop Rows with Null Values
df_mod = df_mod.dropna(subset=['Date'])
print(f"Number of null dates after dropping: {df_mod['Date'].isnull().sum()}")

Number of null dates after dropping: 0


In [21]:
# Step 4: Extract Year, Month, and Day
df_mod['Year'] = df_mod['Date'].dt.year
df_mod['Month'] = df_mod['Date'].dt.month
df_mod['Day'] = df_mod['Date'].dt.day

# Verify the transformations
print(df_mod[['Date', 'Year', 'Month', 'Day']].head(10))

        Date  Year  Month  Day
0 2024-09-16  2024      9   16
1 2024-08-26  2024      8   26
2 2024-08-06  2024      8    6
3 2024-07-23  2024      7   23
4 2024-07-18  2024      7   18
5 2024-07-08  2024      7    8
6 2024-07-05  2024      7    5
7 2024-07-04  2024      7    4
8 2024-07-04  2024      7    4
9 2024-07-04  2024      7    4


### Analysis of the Year Column

In this section, we will analyze the `Year` column of the dataset. We will examine the distribution of years, identify any anomalies, and clean the data as necessary. This analysis will help us understand the temporal trends in the dataset and prepare the data for further analysis and visualization.
```

In [22]:
# Extract unique values from the 'Year' column
unique_years = df_mod['Year'].unique()

# Show the unique years
print(f"Unique years: {unique_years}")

Unique years: [2024 2023 2022 2021 2020 2010 2019 2018 2017 2016 2015 2014 2013 2012
 2011 2008 2009 2006 2007 2005 2004 2003 2002 2001 2000 1999 1998 1997
 1996 1995 1984 1994 1993 1992 1991 1990 1989 1988 1987 1986 1985 1983
 1982 1962 1981 1980 1979 1978 1977 1976 1975 1974 1973 1972 1971 1970
 1969 1968 1967 1966 1965 1964 1954 1963 1961 1960 1959 1958 1957 1956
 1955 1953 1952 1951 1950 1949 1948 1848 1947 1946 1945 1944 1943 1942
 1941 1940 1939 1938 1937 1936 1935 1934 1933 1932 1931 1930 1929 1928
 1927 1926 1925 1924 1923 1922 1921 1920 1919 1918 1917 1916 1915 1914
 1913 1912 1911 1910 1909 1908 1907 1906 1905 1904 1903 1902 1901 1900
 1899 1898 1897 1896 1895 1894 1893 1892 1891 1890 1889 1888 1887 1886
 1885 1884 1883 1882 1881 1880 1879 1878 1877 1876 1875 1874 1873 1872
 1871 1870 1869 1868 1867 1866 1865 1864 1863 1862 1861 1860 1859 1858
 1857 1856 1855 1853 1852 1851 1850 1849 1847 1846 1845 1844 1842 1841
 1840 1839 1837 1836 1835 1834 1832 1831 1830 1829 1828 1827 18

In [23]:
# Show the first few rows of the 'Year' column
print(df_mod['Year'].head(10))

0    2024
1    2024
2    2024
3    2024
4    2024
5    2024
6    2024
7    2024
8    2024
9    2024
Name: Year, dtype: int32


In [24]:
# Count the frequency of each year
year_counts = df_mod['Year'].value_counts().sort_index()

# Display the frequency of each year
print(year_counts)

Year
1691      1
1703      1
1723      1
1733      1
1738      1
       ... 
2020     99
2021    107
2022     98
2023     97
2024     34
Name: count, Length: 237, dtype: int64


In [25]:
# Verify the data type of the 'Year' column
print(df_mod['Year'].dtype)

int32


In [26]:
# Convert the 'Year' column to a string
df_mod['Year'] = df_mod['Year'].astype(str)

# Verify the data type of the 'Year' column
print(df_mod['Year'].dtype)

object



### Analysis of the Type Column

In this section, we will analyze the `Type` column of the dataset. We will examine the different types of incidents recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the nature of the incidents and prepare the data for further analysis and visualization.


In [27]:
# Extract unique values from the 'Type' column
unique_types = df_mod['Type'].unique()
print(f"Unique types: {unique_types}")

Unique types: ['Unprovoked' ' Provoked' 'Provoked' 'Questionable' 'Watercraft'
 'Sea Disaster' nan '?' 'Unconfirmed' 'Unverified' 'Invalid'
 'Under investigation' 'Boat']


In [28]:
# Step 1: 
# Remove leading/trailing spaces
# Replace '?' and 'nan' with 'Unknown'
# Remplace 'Questionable', 'Unverified', 'Invalid', 'Under investigation' with 'Unconfirmed'

df_mod = f.type_column(df_mod)


In [29]:
# Step 2: Verify the cleaned unique values
unique_types_cleaned = df_mod['Type'].unique()
print(f"Cleaned unique types: {unique_types_cleaned}")

Cleaned unique types: ['Unprovoked' 'Provoked' 'Unconfirmed' 'Watercraft' 'Sea Disaster' nan
 'Unknown' 'Boat']


In [30]:
# Step 3: Count the frequency of each type
type_counts_cleaned = df_mod['Type'].value_counts()
print(type_counts_cleaned)

Type
Unprovoked      4820
Provoked         612
Unconfirmed      542
Watercraft       337
Sea Disaster     221
Boat               5
Unknown            1
Name: count, dtype: int64


In [31]:
# Calculate the mode of the 'Type' column
mode_type = df_mod['Type'].mode()[0]

# Replace nan values with the mode
df_mod['Type'].fillna(mode_type, inplace=True)

# Verify that there are no nan values in the 'Type' column
nan_count = df_mod['Type'].isna().sum()
print(f"Number of nan values in 'Type' after imputation: {nan_count}")

Number of nan values in 'Type' after imputation: 0



### Analysis of the Country Column

In this section, we will analyze the `Country` column of the dataset. We will examine the different countries where incidents have been recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the geographical distribution of the incidents and prepare the data for further analysis and visualization.

In [32]:
# Extract unique values from the 'Country' column
unique_countries = df_mod['Country'].unique()
print(f"Unique countries: {unique_countries}")


Unique countries: ['Morocco' 'Jamaica' 'Belize' 'Australia' 'USA' 'Maldive Islands'
 'Turks and Caicos' 'French Polynesia' 'Tobago' 'Bahamas' 'AUSTRALIA'
 'INDIA' 'TRINIDAD' 'BAHAMAS' 'SOUTH AFRICA' 'MEXICO' 'NEW ZEALAND'
 'EGYPT' 'Mexico' 'BELIZE' 'Coral Sea' 'SPAIN' 'PORTUGAL' 'SAMOA'
 'COLOMBIA' 'ECUADOR' 'CUBA' 'BRAZIL' 'SEYCHELLES' 'NEW CALEDONIA'
 'ARGENTINA' 'FIJI' 'MeXICO' 'Maldives' 'South Africa' 'ENGLAND' 'JAPAN'
 'INDONESIA' 'JAMAICA' 'MALDIVES' 'FRENCH POLYNESIA' 'THAILAND' 'COLUMBIA'
 'COSTA RICA' 'New Zealand' 'British Overseas Territory' 'CANADA' 'JORDAN'
 'ST KITTS / NEVIS' 'ST MARTIN' 'PAPUA NEW GUINEA' 'REUNION ISLAND'
 'ISRAEL' 'CHINA' 'IRELAND' 'ITALY' 'MALAYSIA' 'LIBYA' nan 'MAURITIUS'
 'SOLOMON ISLANDS' 'ST HELENA, British overseas territory' 'COMOROS'
 'REUNION' 'UNITED KINGDOM' 'UNITED ARAB EMIRATES' 'PHILIPPINES'
 'CAPE VERDE' 'Fiji' 'DOMINICAN REPUBLIC' 'CAYMAN ISLANDS' 'ARUBA'
 'MOZAMBIQUE' 'PUERTO RICO' 'ATLANTIC OCEAN' 'GREECE' 'ST. MARTIN'
 'FRANCE' 'TRIN

In [33]:
# Convert all country in lower
df_mod = f.country_cleaned(df_mod)

In [34]:
# Extract unique values from the 'Country' column again
unique_countries_cleaned = df_mod['Country'].unique()
print(f"Cleaned unique countries: {unique_countries_cleaned}")


Cleaned unique countries: ['Morocco' 'Jamaica' 'Belize' 'Australia' 'Usa' 'Maldive Islands'
 'Turks And Caicos' 'French Polynesia' 'Tobago' 'Bahamas' 'India'
 'Trinidad' 'South Africa' 'Mexico' 'New Zealand' 'Egypt' 'Coral Sea'
 'Spain' 'Portugal' 'Samoa' 'Colombia' 'Ecuador' 'Cuba' 'Brazil'
 'Seychelles' 'New Caledonia' 'Argentina' 'Fiji' 'Maldives' 'England'
 'Japan' 'Indonesia' 'Thailand' 'Columbia' 'Costa Rica'
 'British Overseas Territory' 'Canada' 'Jordan' 'St Kitts  Nevis'
 'St Martin' 'Papua New Guinea' 'Reunion Island' 'Israel' 'China'
 'Ireland' 'Italy' 'Malaysia' 'Libya' nan 'Mauritius' 'Solomon Islands'
 'St Helena British Overseas Territory' 'Comoros' 'Reunion'
 'United Kingdom' 'United Arab Emirates' 'Philippines' 'Cape Verde'
 'Dominican Republic' 'Cayman Islands' 'Aruba' 'Mozambique' 'Puerto Rico'
 'Atlantic Ocean' 'Greece' 'France' 'Trinidad  Tobago' 'Kiribati'
 'Diego Garcia' 'Taiwan' 'Palestinian Territories' 'Guam' 'Nigeria'
 'Tonga' 'Scotland' 'Croatia' 'Saudi Arab

In [35]:
# Cheking for missing values in the 'Country' column
country_null_count = df_mod['Country'].isnull().sum()
print(f"Number of missing values in 'Country': {country_null_count}")

Number of missing values in 'Country': 43


In [36]:
# Calculate the % of missing values in the 'Country' column
country_null_percent = country_null_count / len(df_mod) * 100
print(f"Percentage of missing values in 'Country': {country_null_percent:.2f}%")


Percentage of missing values in 'Country': 0.66%


In [37]:
# Calculate the mode of the 'Country' column
mode_country = df_mod['Country'].mode()[0]

# Replace missing values with the mode
df_mod['Country'].fillna(mode_country, inplace=True)

# Verify that there are no missing values in the 'Country' column
country_null_count_after = df_mod['Country'].isnull().sum()
print(f"Number of missing values in 'Country' after imputation: {country_null_count_after}")


Number of missing values in 'Country' after imputation: 0


### Analysis of the State Column

In this section, we will analyze the `State` column of the dataset. We will examine the different states recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the geographical distribution of incidents within countries and prepare the data for more detailed analysis and visualization.


In [38]:
# Extract unique values from the 'State' column
unique_states = df_mod['State'].unique()
print(f"Unique states: {unique_states}")

Unique states: ['Southern Morocco' 'Montego Bay' 'Gulf of Honduras' 'NSW'
 'Western Australia' 'Florida' 'Texas' 'Floria' 'California'
 'Maahvah Laamu Atoll' 'Galveston' 'Providenciales Island'
 'Gambier Islands' 'South Carolina' 'Graet Courland Bay'
 'Clarence Town Long Isand' 'South Australia' 'Nunu Atoll' 'Queensland'
 'Hawaii' 'Maharashtra' nan 'New  South Wales' 'Paradise Island'
 'Eastern Cape Province' 'Sonora' 'New South Wales' 'South Island'
 'New Providence   Isoad' 'Jalisco' 'Grand  Bahama Island'
 'Westerm Australia' 'North Island' 'Southern Red Sea' 'Quintana Roo'
 'Belize District' 'South Sinai' 'North Carolina' 'Valencia' 'Norte'
 "Savai'i island" 'New York' 'San Andrés Island' 'Galapagos Islands'
 'Freeport' 'Lucayan Lucayan Archipelago' ' Utah' 'Mayabeque'
 'Easten Cape Province' 'Artemisa' 'Pernambuco' 'Praslin Island'
 'South Province' 'Patagonia' 'Taveuni Island' 'Gulf of California'
 'Vaavu Atoll' 'KNZ' 'New South ales' 'Samoa' 'Louisiana' 'Mississippi'
 'Western  

In [39]:
# Clean the 'State' column using the custom function
df_mod = f.state_cleaned(df_mod)

In [40]:
# Count the number of unique values in the 'State' column
df_mod["State"].nunique() 

822

In [41]:
# Cheking for missing values in the 'State' column
state_null_count = df_mod['State'].isnull().sum()
print(f"Number of missing values in 'State': {state_null_count}")

Number of missing values in 'State': 423


In [42]:
# Calculate the % of missing values in the 'State' column
state_null_percent = state_null_count / len(df_mod) * 100
print(f"Percentage of missing values in 'State': {state_null_percent:.2f}%")


Percentage of missing values in 'State': 6.45%


In [43]:
# Calculate the mode of the 'State' column
mode_state = df_mod['State'].mode()[0]

# Replace missing values with the mode
df_mod['State'].fillna(mode_state, inplace=True)

# Verify that there are no missing values in the 'State' column
state_null_count_after = df_mod['State'].isnull().sum()
print(f"Number of missing values in 'State' after imputation: {state_null_count_after}")


Number of missing values in 'State' after imputation: 0


### Analysis of the Location Column

In this section, we will analyze the `Location` column of the dataset. We will examine the different sex recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the geographical distribution of incidents within countries and prepare the data for more detailed analysis and visualization.

In [44]:
# Show all the unique values in the 'Location' column
for i in df_mod["Location"].unique(): 
    print(i)

West of Dakhla
Falmouth
Central America
North Shore Beach, Port Macquarie
Trigg beach Sterling
Ponce de Leon Inlet Volusia County
New Smyrna Beach
South Padre Island
New Smyrna Beach Beach Access 14 Volusia County
Rosemary Beach Walton County
Del Mar Beach San Diego
Indian Ocean
Jamaica Beach
Leward reef
Mangareva Atoll
Y-73' Ledge reef Chareston Coast
Turtle beach
Flying Fish Marina
West beach Glenelg
Lighthouse bombie surf spot Exemouth
Unspecified
Bargara Beach
Old Man's, Waikiki
Rainbows, Oahu
Sandlnd Island, Jurian Bay
Vaitarna River, Palghar District
nan
Elizabeth Bay, Sydney Harbor
Paradise Island Resort
Walkers Beach, Elliston
Papiesfontein
Baby Beach, Maui
Tojahui Beach, Yavaros, Huatabampo
Ethel Beach
Old Bar
Wedge Island
Riverton
1770 
Sandals Resort
San Patricio Melaque
Clack Island
Tiger Beach
Port Noarlunga
Juno Beach, Palm Beach County
Peak Island
Cape Pallister
Fury Shoal
Streaky Bay
Pua’ena Point,  Haleiwa, Oahu
Banco Chinchorro Biosphere Reserve
Alexander Bay, Esperan

In [45]:
# Count the number of unique values in the 'Location' column
df_mod["Location"].nunique()

4314

In [46]:
# Display the frequency of each location
df_mod["Location"].value_counts()

Location
New Smyrna Beach, Volusia County          189
Daytona Beach, Volusia County              30
Ponce Inlet, Volusia County                28
Myrtle Beach, Horry County                 22
Melbourne Beach, Brevard County            20
                                         ... 
Latitude Reef, near Forster                 1
Motunui                                     1
Surfers' Corner, Muizenberg, False Bay      1
Velzyland                                   1
Below the English fort, Trincomalee         1
Name: count, Length: 4314, dtype: int64

In [47]:
# Cheking for missing values in the 'Location' column
location_null_count = df_mod['Location'].isnull().sum()
print(f"Number of missing values in 'Location': {location_null_count}")
# Calculate the % of missing values in the 'Location' column
location_null_percent = location_null_count / len(df_mod) * 100
print(f"Percentage of missing values in 'Location': {location_null_percent:.2f}%")

Number of missing values in 'Location': 489
Percentage of missing values in 'Location': 7.46%


In [48]:
# Convert all location in lower
df_mod = f.location_cleaned(df_mod)

# Apply the cleaning function to the 'Location' column
df_mod = f.clean_location_column(df_mod, 'Location')

# Verify the changes
print(df_mod['Location'].unique())

['West Of Dakhla' 'Falmouth' 'Central America' ... 'Ocracoke Inlet'
 'Panama Bay 8ºn, 79ºw' 'Below The English Fort, Trincomalee']


In [49]:
# Sustitute the missing values with the mode
mode_location = df_mod['Location'].mode()[0]
df_mod['Location'].fillna(mode_location, inplace=True)

# Verify that there are no missing values in the 'Location' column
location_null_count_after = df_mod['Location'].isnull().sum()
print(f"Number of missing values in 'Location' after imputation: {location_null_count_after}")


Number of missing values in 'Location' after imputation: 0


### Analysis of the Activity Column

In this section, we will analyze the `Activity` column of the dataset. We will examine the different sex recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the geographical distribution of incidents within countries and prepare the data for more detailed analysis and visualization.

In [50]:
# Count the number of unique values in the 'Activity' column
df_mod["Activity"].nunique()

1485

In [51]:
# Display all the unique values in the 'Activity' column
for values in df_mod["Activity"].unique(): 
    print(values)

Swimming - jumped off yacht
Spearfishing
Swimming - Diving 
Surfing
Diving into Water
Wading
Swimming
Playing football in the water
Swimming ocean training 
Military ocean training
snorkelling
Freediving
Jumpinf from dock to boat acidentally fell into water at marina
Fishing
nan
Free diving
Foil Boarding
Snorkeling
Paddle boarding
Scuba diving
On a round-the-world expedition
Kayak fishing
Towing a dead whale out to sea
Boogie boarding
Sitting
Watching the sardine run
Standing
Jumped overboard
Shark Feeding
Diving for crayfish
Windsurfing
Kite-Surfing
Surf-sking
Diving with  surface-supplied air
Spearishing
Snorkeling at Shark Feeding site
Diving/Shark Feeding
Diving
Surf skiing 
Shipwreck
Swimming  or Snorkeling
Kayaking
Paddleboarding
Lobstering
Lifeguard Training Exercise
Lifeguard Exercises
Scalloping
Jumped into water
Playing in the water
Feeding sharks & stingrays
Fihing
Surface Supplied  Diving
Floating in inflatable pool ring
Fishing for sharks
Wing Foil Surfing
Swimmingq
Fishin

In [52]:
# Display the frequency of each activity
df_mod["Activity"].value_counts()

Activity
Surfing                                 1105
Swimming                                 938
Fishing                                  468
Spearfishing                             356
Wading                                   169
                                        ... 
Hookah Diving                              1
Sinking of the ferryboat Bongbong 1        1
Sinking of the ferryboat Don Juan          1
Resting on body board                      1
Wreck of  large double sailing canoe       1
Name: count, Length: 1485, dtype: int64

In [53]:
# Convert all Activity in lower
df_mod = f.activity_cleaned(df_mod)

# Apply the cleaning function to the 'Activity' column
df_mod = f.clean_activity_column(df_mod, 'Activity')

# Verify the changes
print(df_mod['Activity'].unique())

['Swimming  Jumped Off Yacht' 'Spearfishing' 'Swimming  Diving' ...
 'Crew Swimming Alongside Their Anchored Ship' '4 Men Were Bathing'
 'Wreck Of  Large Double Sailing Canoe']


In [54]:
# Cheking for missing values in the 'Activity' column
activity_null_count = df_mod['Activity'].isnull().sum()
print(f"Number of missing values in 'Activity': {activity_null_count}")

Number of missing values in 'Activity': 508


In [55]:
# Calculate the % of missing values in the 'Activity' column
activity_null_percent = activity_null_count / len(df_mod) * 100
print(f"Percentage of missing values in 'Activity': {activity_null_percent:.2f}%")

Percentage of missing values in 'Activity': 7.75%


In [56]:
# Calculate the mode of the 'Activity' column
mode_activity = df_mod['Activity'].mode()[0]

# Replace missing values with the mode
df_mod['Activity'].fillna(mode_activity, inplace=True)

# Verify that there are no missing values in the 'Activity' column
activity_null_count_after = df_mod['Activity'].isnull().sum()
print(f"Number of missing values in 'Activity' after imputation: {activity_null_count_after}")


Number of missing values in 'Activity' after imputation: 0


### Analysis of the Name Column

In this section, we will analyze the `Name` column of the dataset. We will examine the different sex recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the geographical distribution of incidents within countries and prepare the data for more detailed analysis and visualization.

In [57]:
# Display the unique values in the 'Name' column
df_mod["Name"]

0              German Tourist
1                Jahmari Reid
2           Annabelle Carlson
3                Kai McKenzie
4                Ryan Lowther
                ...          
6964                     male
6965                    Ahmun
6966    Coast Guard personnel
6967          Jules Patterson
6968                     male
Name: Name, Length: 6554, dtype: object

In [58]:
# Display the unique values in the 'Name' column
df_mod["Name"].unique()

array(['German Tourist', 'Jahmari Reid', 'Annabelle Carlson', ...,
       'Mr. Masury', 'Ahmun', 'Jules Patterson'], dtype=object)

In [59]:
# Count the number of missing values in the 'Name' column
df_mod["Name"].isna().sum()

202

In [60]:
# Count the frequency of each unique value in the 'Name' column
df_mod["Name"].value_counts()

Name
male                           610
female                         128
boy                             16
2 males                         15
child                           11
                              ... 
Michaela Grogan                  1
Bernard 'Butch' Connor, Jr.      1
Walmir Pereira da Silva          1
Ryan Eckstrum                    1
Jules Patterson                  1
Name: count, Length: 5435, dtype: int64

### Analysis of the Sex Column

In this section, we will analyze the `Sex` column of the dataset. We will examine the different sex recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the geographical distribution of incidents within countries and prepare the data for more detailed analysis and visualization.

In [61]:
# Display the unique values in the 'Sex' column
df_mod["Sex"].unique()

array(['F', 'M', nan, ' M', 'M ', 'lli', 'M x 2', 'N', '.'], dtype=object)

In [62]:
# Clean the 'Sex' column using the custom function
df_mod = f.sex_clean(df_mod)

In [63]:
# Display the unique values in the 'Sex' column
df_mod['Sex'].unique()

array(['F', 'M', nan, 'undefined'], dtype=object)

In [64]:
# Cheking for undefined values in the 'Sex' column
sex_und_count = len(df_mod[df_mod['Sex'] == 'undefined'])
print(f"Number of missing values in 'Sex': {sex_und_count}")
# Calculate the % of missing values in the 'Sex' column
sex_und_percent = sex_und_count / len(df_mod) * 100
print(f"Percentage of undefined values in 'Sex': {sex_und_percent:.2f}%")

Number of missing values in 'Sex': 4
Percentage of undefined values in 'Sex': 0.06%


In [65]:
# Cheking for missing values in the 'Sex' column
sex_null_count = df_mod['Sex'].isnull().sum()
print(f"Number of missing values in 'Sex': {sex_null_count}")
# Calculate the % of missing values in the 'Sex' column
sex_null_percent = sex_null_count / len(df_mod) * 100
print(f"Percentage of missing values in 'Sex': {sex_null_percent:.2f}%")


Number of missing values in 'Sex': 528
Percentage of missing values in 'Sex': 8.06%


In [66]:
# The correct option for our analyse is replace Nan with 'undefined'

df_mod.fillna('undefined', inplace=True)

C:\Users\Carlota\AppData\Local\Temp\ipykernel_11340\3351348298.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'undefined' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df_mod.fillna('undefined', inplace=True)


In [67]:
# Verify that there are no missing values in the 'Sex' column
sex_null_count_after = df_mod['Sex'].isnull().sum()
print(f"Number of missing values in 'Sex' after imputation: {sex_null_count_after}")

Number of missing values in 'Sex' after imputation: 0


In [68]:
# Cheking for undefined values in the 'Sex' column after cleaning
sex_und_count_after = len(df_mod[df_mod['Sex'] == 'undefined'])
print(f"Number of 'undefined' values in 'Sex': {sex_und_count_after}")
# Calculate the % of missing values in the 'Sex' column
sex_und_percent_after = sex_und_count_after / len(df_mod) * 100
print(f"Percentage of 'undefined' values in 'Sex': {sex_und_percent_after:.2f}%")

Number of 'undefined' values in 'Sex': 532
Percentage of 'undefined' values in 'Sex': 8.12%


In [69]:
# The best option for our analysis is eliminate this values 

df_mod = df_mod[df_mod['Sex'] != 'undefined']

In [70]:
df_mod['Sex'].unique()

array(['F', 'M'], dtype=object)

In [71]:
df_mod['Sex'].value_counts()

Sex
M    5271
F     751
Name: count, dtype: int64

### Analysis of the Age Column

In this section, we will analyze the `Age` column of the dataset. We will examine the different sex recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the geographical distribution of incidents within countries and prepare the data for more detailed analysis and visualization.

In [72]:
#Step 1: Visualize Age Data
df_mod["Age"].unique() #pasar a numero o rango para poder hacer media de edad 

array([30, 16, 15, 23, 41, 14, 26, 'Middle age', 18, 21, 'undefined', 45,
       46, 19, 27, 20, 65, 64, 24, '?', '13', '11', '46', '32', '29',
       '10', '64', '62', '39', '22', '15', '16', '21', '20', '44', '26',
       '47', '66', '55', '30', '59', '8', '50', '52', '45', '34', '38',
       '30s', '37', '19', '20/30', '35', '65', '20s', '77', '60', 38,
       '49', '42', '!2', '73', '18', '24', '50s', '58', '67', '17', '14',
       '6', '41', '53', '68', '43', '40', '51', '31', 39, 58, 51, 17, 10,
       13, 33, 40, 49, 60, 28, '40s', 68, 35, 62, 'teen', 8, 22, 32, 56,
       'Teen', 12, 42, 36, 37, 50, 'M', 9, 57, 34, 25, '!6', 31, '!!', 47,
       55, 11, 7, 71, 48, 59, 53, 54, 75, '45 and 15', 61, 73, 52, 29, 70,
       4, 63, 44, '28 & 22', '60s', "20's", 43, 67, 74, '9 & 60',
       'a minor', 6, 3, 69, 82, 66, 72, '23', '12', '9', '36', '63', '71',
       '48', '70', '18 months', '57', '7', '28', '33', '61', '25', '74',
       '27', '3', '56', '28 & 26', '5', '54', '86', '18 

In [73]:
# Step 2: Clean

df_mod = f.clean_age(df_mod)

In [74]:
df_mod['Age'].unique()

array([nan, 13., 11., 46., 32., 29., 10., 64., 62., 39., 22., 15., 16.,
       21., 20., 44., 26., 47., 66., 55., 30., 59.,  8., 50., 52., 45.,
       34., 38., 37., 19., 35., 65., 77., 60., 49., 42., 73., 18., 24.,
       58., 67., 17., 14.,  6., 41., 53., 68., 43., 40., 51., 31., 23.,
       12.,  9., 36., 63., 71., 48., 70., 57.,  7., 28., 33., 61., 25.,
       74., 27.,  3., 56.,  5., 54., 86., 87.,  4., 81., 78.,  1.])

In [75]:
# Step 3: null values
# Number of missing values in Age column
nan_count_age = df_mod['Age'].isna().sum()
nan_count_age

4276

In [76]:
# Calculate the % of missing values in the 'Age' column
age_null_percent = nan_count_age / len(df_mod) * 100
print(f"Percentage of missing values in 'Age': {age_null_percent:.2f}%")

Percentage of missing values in 'Age': 71.01%


In [77]:
# debido al gran porcentaje, sustituimos por la media 
# Calculate the mean of the 'Type' column
mean_age = df_mod['Age'].mean().astype(int)
mean_age



# Replace nan values with the mode
df_mod['Age'].fillna(mean_age, inplace=True)

# Verify that there are no nan values in the 'Type' column
nan_count = df_mod['Age'].isna().sum()
print(f"Number of nan values in 'Age' after imputation: {nan_count}")

Number of nan values in 'Age' after imputation: 0


In [78]:
# Step 4: Change data type to integer
df_mod['Age'] = df_mod['Age'].astype(int)
df_mod['Age'].unique()

array([28, 13, 11, 46, 32, 29, 10, 64, 62, 39, 22, 15, 16, 21, 20, 44, 26,
       47, 66, 55, 30, 59,  8, 50, 52, 45, 34, 38, 37, 19, 35, 65, 77, 60,
       49, 42, 73, 18, 24, 58, 67, 17, 14,  6, 41, 53, 68, 43, 40, 51, 31,
       23, 12,  9, 36, 63, 71, 48, 70, 57,  7, 33, 61, 25, 74, 27,  3, 56,
        5, 54, 86, 87,  4, 81, 78,  1])

In [79]:
# Step 5: Value counts

df_mod["Age"].value_counts()

Age
28    4310
17      82
15      77
20      74
18      71
      ... 
87       1
4        1
81       1
78       1
1        1
Name: count, Length: 76, dtype: int64

### Analysis of the Injury Column

In this section, we will analyze the `Injury` column of the dataset. We will examine the different injuries recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the geographical distribution of incidents within countries and prepare the data for more detailed analysis and visualization.

In [80]:
# Display all unique values in the 'Injury' column
for i in df_mod["Injury"].unique():
    print(i)

Leg severed
Head right arm and leg severed
Right lower leg amputated after serious bite
Serious injury to leg
Minor injury to lower left leg
Lower left leg injury
Minor injury to left foot
Bite to left leg calf muscle removed
Minor cuts to calf
Superficial wounds to leg
Minor injury to right foot four tendons ruptured
Bite to left leg  
Significant injuries to pelvic and abdominal region
Significant bites to torso left arm and hand
Wounds to lower extremities Fatal
Bitten on left hand by 4-5ft shark
Major damage to right leg later amputated lower left leg left hand and forearm
Bite left forearm several ripped tendons severed nerves deep muscle wounds
Bitten on left arm
Bite wounds to left arm and leg wounds to abdomen and left hand
Bitten on left leg puncture wounds to kneecap
Minor bite to left leg
Minor injuries left foot
Minor injury to left leg
Minor injuries to back, abdomen and legs
No injury, shark bit surfboard
Lacerations to left foot
Leg bitten
Calf of lower left leg injured


In [81]:
df_mod["Injury"].nunique() #revisar datos validos

3539

In [82]:
# Cheking for missing values in the 'Injury' column
injury_null_count = df_mod['Injury'].isnull().sum()
print(f"Number of missing values in 'Injury': {injury_null_count}")
# Calculate the % of missing values in the 'Injury' column
injury_null_percent = injury_null_count / len(df_mod) * 100
print(f"Percentage of missing values in 'Injury': {injury_null_percent:.2f}%")

Number of missing values in 'Injury': 0
Percentage of missing values in 'Injury': 0.00%


In [83]:
df_mod["Injury"].value_counts()

Injury
FATAL                                                                                            749
Foot bitten                                                                                       96
Leg bitten                                                                                        73
No injury                                                                                         72
Survived                                                                                          71
                                                                                                ... 
2 lacerations on leg, each 4" to 5" long                                                           1
Cut on finger, board bitten                                                                        1
Lacerations to hands & right leg when he tried to pull shark from the water PROVOKED INCIDENT      1
Minor cuts to dorsum & sole of left foot when he stepped on shark PROVOKED INCIDENT 

In [84]:
# Convert all country in lower
df_mod = f.injury_cleaned(df_mod)

# Apply the cleaning function to the 'Injury' column
df_mod = f.clean_injury_column(df_mod, 'Injury')

# Verify the changes
print(df_mod['Injury'].unique())

['Leg severed' 'Head right arm and leg severed'
 'Right lower leg amputated after serious bite' ...
 'Fatal, leg stripped of flesh'
 'Fatal, knocked overboard by tail of shark  carried off by shark'
 'Fatal shark bit him in half, carrying away the lower extremities']


12


### Analysis of the Species Column

In this section, we will analyze the `Species` column of the dataset. We will examine the different species recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the types of sharks involved in the incidents and prepare the data for further analysis and visualization.


In [85]:
# Display all the unique values in the 'Species' column
for i in df_mod["Species"].unique():
    print(i)

Reportedly a Great White
Reportedly Tiger Shark
Reef shark 6ft (1.8m)
3m White Shark
Undetermined small shall shark
4-5ft Blacktip shark
Not specified
Bull shark 6 ft
6 ft shark
6ft shark
Not specified Bull shark highly suspect
Large shark
2.5m shark
3m Sandbar shark
Bull shark 3m 
Carribean reef shark 7ft
Great White
Tiger shark
Tiger shark 8'
3' to 4' shark
Bull shark, 7'
Blacktip reef shark
Bull shark
undefined
White shark
Raggedtooth shark
White shark, 13'
Broze whaler?
Sevengill shark
1.8m bull shark
Lemon shark
Oceanic whitetip shark
White Shark, 4.5 m
8' tiger shark
Bronze whaler, 6'
Nurse shark, 5' to 6'
10' tiger shark
10' to 12' Galapagos shark
Tiger shark?
Cookiecutter sharks
White shark, 3.8-4.2m
Blue shark
Wfite shark
Wobbegong shark
8' shark
4' shark
3' shark
Sand tiger shark
Caribbean rreef shark
Shark involvement not confirmes
18" - 23" Horn shark
Bull shark, 4'
1m shark
Bull shark, 6'
Bronze whaler shark, 1.5 m
Tiger shark, 8'-10'
Blacktip reef shark 5'
Authorities rep

In [86]:
# Display the number of unique values in the 'Species' column
df_mod["Species"].nunique()

1523

In [87]:
# Display the frequency of each unique value in the 'Species' column
df_mod["Species"].value_counts()

Species
undefined                                                                                                                        2659
White shark                                                                                                                       162
Shark involvement prior to death was not confirmed                                                                                 83
Tiger shark                                                                                                                        80
Invalid                                                                                                                            79
                                                                                                                                 ... 
Caribbean reef sharks                                                                                                               1
Tiger shark, less than 10'                            

In [88]:
# Display the number of missing values in the 'Species' column
species_null_count = df_mod['Species'].isnull().sum()
print(f"Number of missing values in 'Species': {species_null_count}")

Number of missing values in 'Species': 0


In [89]:
# Aplicar la función de limpieza y normalización a la columna 'Species'
df_mod = f.clean_and_normalize_species(df_mod, 'Species')

# Verificar los cambios
print(df_mod['Species'].value_counts().head(12))

Species
undefined            2659
White Shark           633
Tiger Shark           301
Bull Shark            199
Blacktip Shark        127
Nurse Shark           102
Bronze Shark           64
Blue Shark             51
Mako Shark             44
Raggedtooth Shark      42
Lemon Shark            42
Hammerhead Shark       41
Name: count, dtype: int64



### Adding an Ocean Column

In this section, we will add a new column to the DataFrame that indicates the ocean associated with each country. This will help us analyze the geographical distribution of incidents in relation to different oceans. We will use a mapping function to assign the appropriate ocean to each country based on its location.


In [90]:
# Display unique values in the 'Country' column
df_mod['Country'].unique()

array(['morocco', 'jamaica', 'belize', 'australia', 'usa',
       'maldive islands', 'turks and caicos', 'french polynesia',
       'tobago', 'bahamas', 'india', 'trinidad', 'south africa', 'mexico',
       'new zealand', 'egypt', 'coral sea', 'spain', 'portugal', 'samoa',
       'colombia', 'ecuador', 'cuba', 'brazil', 'seychelles',
       'new caledonia', 'argentina', 'fiji', 'maldives', 'england',
       'japan', 'indonesia', 'thailand', 'columbia', 'costa rica',
       'british overseas territory', 'canada', 'jordan',
       'st kitts  nevis', 'st martin', 'papua new guinea',
       'reunion island', 'israel', 'china', 'ireland', 'italy',
       'malaysia', 'mauritius', 'solomon islands',
       'st helena british overseas territory', 'reunion',
       'united kingdom', 'united arab emirates', 'philippines',
       'dominican republic', 'aruba', 'mozambique', 'puerto rico',
       'france', 'trinidad  tobago', 'kiribati', 'diego garcia', 'taiwan',
       'palestinian territories', 

In [91]:
# Añadir la columna de océanos y mares
df_mod = f.add_oceans_column(df_mod, 'Country', 'Ocean_Sea')

# Verificar los cambios
print(df_mod[['Country', 'Ocean_Sea']].head(20))

Valores únicos de la columna de países después de convertir a minúsculas y eliminar espacios:
['morocco' 'jamaica' 'belize' 'australia' 'usa' 'maldive islands'
 'turks and caicos' 'french polynesia' 'tobago' 'bahamas' 'india'
 'trinidad' 'south africa' 'mexico' 'new zealand' 'egypt' 'coral sea'
 'spain' 'portugal' 'samoa' 'colombia' 'ecuador' 'cuba' 'brazil'
 'seychelles' 'new caledonia' 'argentina' 'fiji' 'maldives' 'england'
 'japan' 'indonesia' 'thailand' 'columbia' 'costa rica'
 'british overseas territory' 'canada' 'jordan' 'st kitts  nevis'
 'st martin' 'papua new guinea' 'reunion island' 'israel' 'china'
 'ireland' 'italy' 'malaysia' 'mauritius' 'solomon islands'
 'st helena british overseas territory' 'reunion' 'united kingdom'
 'united arab emirates' 'philippines' 'dominican republic' 'aruba'
 'mozambique' 'puerto rico' 'france' 'trinidad  tobago' 'kiribati'
 'diego garcia' 'taiwan' 'palestinian territories' 'guam' 'nigeria'
 'tonga' 'scotland' 'saudi arabia' 'chile' 'kenya' '

In [92]:
# Imprime los 10 primeros valores de las columnas country y ocean_sea
print(df_mod[['Country', 'Ocean_Sea']].head(10))

     Country                         Ocean_Sea
0    morocco                    Atlantic Ocean
1    jamaica                     Caribbean Sea
2     belize                     Caribbean Sea
3  australia    Indian Ocean and Pacific Ocean
4  australia    Indian Ocean and Pacific Ocean
5        usa  Atlantic Ocean and Pacific Ocean
6        usa  Atlantic Ocean and Pacific Ocean
7        usa  Atlantic Ocean and Pacific Ocean
8        usa  Atlantic Ocean and Pacific Ocean
9        usa  Atlantic Ocean and Pacific Ocean


In [94]:
# Imprime la frecuencia de los valores de la columna ocean_sea
df_mod['Ocean_Sea'].value_counts()

Ocean_Sea
Atlantic Ocean and Pacific Ocean                   2122
Indian Ocean and Pacific Ocean                     1176
Pacific Ocean                                       541
Atlantic Ocean and Indian Ocean                     460
Atlantic Ocean                                      284
Indian Ocean                                        131
Caribbean Sea                                        85
Pacific Ocean and Gulf of Mexico                     79
Mediterranean Sea                                    62
Pacific Ocean and Caribbean Sea                      50
Red Sea                                              48
Atlantic Ocean and Mediterranean Sea                 38
Adriatic Sea                                         23
South China Sea                                      14
Persian Gulf and Caspian Sea                         14
Atlantic Ocean, Pacific Ocean, and Arctic Ocean      12
Indian Ocean and Andaman Sea                         11
Mediterranean Sea and Black Sea       

In [93]:
# Display the isnull sum of the 'Ocean_Sea' column
df_mod['Ocean_Sea'].isnull().sum()


237

In [94]:
# Eliminar los valores nulos de la columna 'Ocean_Sea'
df_mod = df_mod.dropna(subset=['Ocean_Sea'])

# Verificar los cambios
df_mod['Ocean_Sea'].isnull().sum()

0

In [95]:
df_mod["Unnamed: 11"]

0       Y
1       Y
2       N
3       N
4       N
       ..
6963    Y
6964    Y
6965    Y
6966    Y
6967    Y
Name: Unnamed: 11, Length: 5178, dtype: object

In [96]:
df_mod["Unnamed: 11"].value_counts()

Unnamed: 11
N            3929
Y            1118
undefined      90
UNKNOWN        27
 N              5
F               3
M               2
n               1
Nq              1
2017            1
Y x 2           1
Name: count, dtype: int64

In [97]:
df_mod.drop(columns=["Unnamed: 11"], inplace=True)
df_mod.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5178 entries, 0 to 6967
Data columns (total 18 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Date            5178 non-null   datetime64[ns]
 1   Year            5178 non-null   object        
 2   Type            5178 non-null   object        
 3   Country         5178 non-null   object        
 4   State           5178 non-null   object        
 5   Location        5178 non-null   object        
 6   Activity        5178 non-null   object        
 7   Name            5178 non-null   object        
 8   Sex             5178 non-null   object        
 9   Age             5178 non-null   int32         
 10  Injury          5178 non-null   object        
 11  Time            5178 non-null   object        
 12  Species         5178 non-null   object        
 13  Source          5178 non-null   object        
 14  original order  5178 non-null   object        
 15  Month    

### Analysis of the Time Column

In this section, we will analyze the `Time` column of the dataset. We will examine the different times recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the geographical distribution of incidents within countries and prepare the data for more detailed analysis and visualization.

In [98]:
df_mod["Time"]

0          1600hr
1       undefined
2         11.30hr
3          1100hr
4          1735hr
          ...    
6963    undefined
6964    undefined
6965    undefined
6966    undefined
6967    undefined
Name: Time, Length: 5178, dtype: object

In [99]:
df_mod["Time"].value_counts()

Time
undefined        2266
Afternoon         172
11h00             114
Morning           113
15h00             112
                 ... 
09h00 - 09h30       1
0830                1
15h56               1
07h56               1
19h00-20h00         1
Name: count, Length: 389, dtype: int64

In [100]:
df_mod["Time"].unique() #mismo formato horaa (plantear rango) y reagrupar los no conocidos

array(['1600hr', 'undefined', '11.30hr', '1100hr', '1735hr', '11hr15',
       '16hr15', '?', '13h15', '9h', 1300, '14h', '15h30', 'Not stated',
       '13h30', '9h15', 'Not advised', '13h40', '12h30', '16h00', '11h30',
       '06h30', '20h00', '13h00', '11h12', '16h30', '15h00', '02h00',
       '16h32', '11h00', 'Morning', '10h30', '13h20', '14h00', '09h00',
       '10h20', '15h05', '17h00', '15h45', '07h45', '10h40', '07h50',
       '10h00', 'Afternoon', '19h30', 'Evening', '17h50', '09h30',
       '08h45', '"Midday"', '13h50', '17h20', '13h45', '10h10', 'Night',
       '1500h ', '19h15', '11h20', '07h15', '07h00', '18h00', '08h00',
       '14h20', '17h30', '07h20', '14h50', '-16h30', '12h00', '17h17',
       '11h15', '19h00', '07h53', '16h10', '11h17', 'Early  morning',
       '13h12', '07h30', '11hoo', '11h43', '10h15', '14h09', '12h15',
       '19h12', '15h20', '16h40', '11h24', '12h50', '07h31', '14h45',
       '19h20', 'Dusk', '11h45', '06h40', '`17h00', '07h51', '11h46',
       

In [101]:
#Step 1: Standarized data
df_mod = f.cleaned_time(df_mod)

In [102]:
df_mod["Time"].unique()

array(['Afternoon', nan, 'Morning', 'Night'], dtype=object)

In [103]:
# Step 2
# Cheking for missing values in the 'Time' column
time_null_count = df_mod['Time'].isnull().sum()
print(f"Number of missing values in 'Time': {time_null_count}")
# Calculate the % of missing values in the 'Time' column
time_null_percent = time_null_count / len(df_mod) * 100
print(f"Percentage of missing values in 'Time': {time_null_percent:.2f}%")

Number of missing values in 'Time': 2326
Percentage of missing values in 'Time': 44.92%


In [104]:
# Step 3: change the missing values to mode
# Calculate the mode of the 'Time' column
mode_time = df_mod['Time'].mode()[0]

# Replace nan values with the mode
df_mod['Time'].fillna(mode_time, inplace=True)

# Verify that there are no nan values in the 'Type' column
nan_count_time = df_mod['Time'].isna().sum()
print(f"Number of nan values in 'Time' after imputation: {nan_count_time}")

Number of nan values in 'Time' after imputation: 0


In [105]:
df_mod['Time'].value_counts()

Time
Afternoon    3670
Morning      1180
Night         328
Name: count, dtype: int64

### Todas las columnas siguientes han sido **eliminadas**

In [69]:
pf["Source"]

0              Andy Currie: Moroccan World News: Sky News
1        Todd Smith: Daily Mail UK: Sky News: People .com
2       Kevin McMurray Tracking Sharks.com New York Po...
3               Nine.com.au: ABC.Net.Au Johannes Marchand
4               Daily Mail: Sky News: The West Australian
                              ...                        
6965                     H. Taunton; N. Bartlett,  p. 234
6966                H. Taunton; N. Bartlett,  pp. 233-234
6967                 F. Schwartz, p.23; C. Creswell, GSAF
6968                                  The Sun, 10/20/1938
6969                                           S.W. Baker
Name: Source, Length: 6970, dtype: object

In [70]:
pf["Source"].unique() #posible a eliminar

array(['Andy Currie: Moroccan World News: Sky News',
       'Todd Smith: Daily Mail UK: Sky News: People .com',
       'Kevin McMurray Tracking Sharks.com New York Post: ABC News: The Independent:',
       ..., 'F. Schwartz, p.23; C. Creswell, GSAF', 'The Sun, 10/20/1938',
       'S.W. Baker'], dtype=object)

In [71]:
pf["Source"].nunique()

5326

- pdf                171
- href formula       176
- href               174
- Case Number        172
- Case Number.1      173
- original order     171
- Unnamed: 21       6969
- Unnamed: 22       6968

In [72]:
pf["pdf"] #eliminar

0                                  NaN
1                                  NaN
2                                  NaN
3                                  NaN
4                                  NaN
                     ...              
6965            ND-0005-RoebuckBay.pdf
6966                 ND-0004-Ahmun.pdf
6967    ND-0003-Ocracoke_1900-1905.pdf
6968        ND-0002-JulesPatterson.pdf
6969                ND-0001-Ceylon.pdf
Name: pdf, Length: 6970, dtype: object

In [10]:
pf_mod.drop(columns=["pdf"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 22 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  href formula    6794 non-null   object 
 16  href            6796 non-null   object 
 17  Case Number     6798 non-null   o

In [73]:
pf["href"] #eliminar

0                                                     NaN
1                                                     NaN
2                                                     NaN
3                                                     NaN
4                                                     NaN
                              ...                        
6965    http://sharkattackfile.net/spreadsheets/pdf_di...
6966    http://sharkattackfile.net/spreadsheets/pdf_di...
6967    http://sharkattackfile.net/spreadsheets/pdf_di...
6968    http://sharkattackfile.net/spreadsheets/pdf_di...
6969    http://sharkattackfile.net/spreadsheets/pdf_di...
Name: href, Length: 6970, dtype: object

In [11]:
pf_mod.drop(columns=["href"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  href formula    6794 non-null   object 
 16  Case Number     6798 non-null   object 
 17  Case Number.1   6797 non-null   o

In [75]:
pf["href formula"] #eliminar

0                                                     NaN
1                                                     NaN
2                                                     NaN
3                                                     NaN
4                                                     NaN
                              ...                        
6965    http://sharkattackfile.net/spreadsheets/pdf_di...
6966    http://sharkattackfile.net/spreadsheets/pdf_di...
6967    http://sharkattackfile.net/spreadsheets/pdf_di...
6968    http://sharkattackfile.net/spreadsheets/pdf_di...
6969    http://sharkattackfile.net/spreadsheets/pdf_di...
Name: href formula, Length: 6970, dtype: object

In [12]:
pf_mod.drop(columns=["href formula"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  Case Number     6798 non-null   object 
 16  Case Number.1   6797 non-null   object 
 17  original order  6799 non-null   f

In [76]:
pf["Case Number"] #eliminar

0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
         ...   
6965    ND.0005
6966    ND.0004
6967    ND.0003
6968    ND.0002
6969    ND.0001
Name: Case Number, Length: 6970, dtype: object

In [13]:
pf_mod.drop(columns=["Case Number"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  Case Number.1   6797 non-null   object 
 16  original order  6799 non-null   float64
 17  Unnamed: 21     1 non-null      o

In [80]:
pf["original order"].unique() #que es???''

array([      nan, 6.802e+03, 6.801e+03, ..., 4.000e+00, 3.000e+00,
       2.000e+00])

In [81]:
pf["Case Number.1"] #eliminar

0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
         ...   
6965    ND.0005
6966    ND.0004
6967    ND.0003
6968    ND.0002
6969    ND.0001
Name: Case Number.1, Length: 6970, dtype: object

In [14]:
pf_mod.drop(columns=["Case Number.1"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 18 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  original order  6799 non-null   float64
 16  Unnamed: 21     1 non-null      object 
 17  Unnamed: 22     2 non-null      o

In [82]:
pf["Unnamed: 21"] #eliminar

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
6965    NaN
6966    NaN
6967    NaN
6968    NaN
6969    NaN
Name: Unnamed: 21, Length: 6970, dtype: object

In [15]:
pf_mod.drop(columns=["Unnamed: 21"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  original order  6799 non-null   float64
 16  Unnamed: 22     2 non-null      object 
dtypes: float64(2), object(15)
memory 

In [83]:
pf["Unnamed: 22"] #eliminar

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
6965    NaN
6966    NaN
6967    NaN
6968    NaN
6969    NaN
Name: Unnamed: 22, Length: 6970, dtype: object

In [16]:
pf_mod.drop(columns=["Unnamed: 22"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  original order  6799 non-null   float64
dtypes: float64(2), object(14)
memory usage: 871.4+ KB


### Save cleaned data

In [115]:
df_mod.to_csv('cleaned_data.csv', index=False)